pkg/sbproxy/reconcile/reconcile_visibilities.go

```
@@ -0,0 +1,351 @@
+	logger := log.C(r.ctx)
+	logger.Debug("ReconciliationTask reconsiling platform and SM visibilities...")
+
+	platformMap := r.convertVisListToMap(platformVis)
+	visibilitiesToCreate := make([]*platform.ServiceVisibilityEntity, 0)
+	for _, visibility := range smVis {
+		key := r.getVisibilityKey(visibility)
+		existingVis := platformMap[key]
+		delete(platformMap, key)
+		if existingVis == nil {
+			visibilitiesToCreate = append(visibilitiesToCreate, visibility)
+		}
+	}
+
+	logger.Debugf("ReconciliationTask %d visibilities will be removed from the platform", len(platformMap))
+	errorOccured := r.deleteVisibilities(platformMap)
+
+	logger.Debugf("ReconciliationTask %d visibilities will be created in the platform", len(visibilitiesToCreate))
+	errorOccured = r.createVisibilities(visibilitiesToCreate) || errorOccured
+
+	return errorOccured
+}
+
+// deleteVisibilities deletes visibilities from platform. Returns true if error has occurred
+func (r ReconciliationTask) deleteVisibilities(visibilities map[string]*platform.ServiceVisibilityEntity) bool {
+	var errorOccured int32 // will be used as atomic bool
+	var wg sync.WaitGroup
+
+	for _, visibility := range visibilities {
+		execAsync(&wg, visibility, &errorOccured, r.deleteVisibility)
+	}
+	return await(&wg, &errorOccured)
+
+}
+
+// createVisibilities creates visibilities from platform. Returns true if error has occurred
+func (r ReconciliationTask) createVisibilities(visibilities []*platform.ServiceVisibilityEntity) bool {
+	var errorOccured int32 // will be used as atomic bool
+	var wg sync.WaitGroup
+
+	for _, visibility := range visibilities {
+		execAsync(&wg, visibility, &errorOccured, r.createVisibility)
+	}
+	return await(&wg, &errorOccured)
+}
+
+func execAsync(wg *sync.WaitGroup, visibility *platform.ServiceVisibilityEntity, errorOccured *int32, f func(*platform.ServiceVisibilityEntity) error) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := f(visibility); err != nil {
```

error is swallowed. it would be best to be propagated up the stack. pass an error channel on which to report in case an error occurs

---

Here, only the first error will be logged so that the logs don't get flooded.